File: code/系統程式/textdata/git0/git0.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>

#define GIT0_DIR ".git0"
#define OBJECTS_DIR GIT0_DIR "/objects"
#define INDEX_FILE GIT0_DIR "/index"
#define COMMITS_FILE GIT0_DIR "/commits"
#define HASH_LEN 8
#define FNV1A_OFFSET 0x811C9DC5UL
#define FNV1A_PRIME 0x01000193UL

// --- 修復 1：在這裡先宣告函數原型 ---
char *strrstr(const char *haystack, const char *needle);

// 輔助函數：檢查文件是否存在
int file_exists(const char *filename) {
    struct stat buffer;
    return (stat(filename, &buffer) == 0);
}

// 輔助函數：檢查目錄是否存在
int dir_exists(const char *dirname) {
    struct stat buffer;
    if (stat(dirname, &buffer) != 0) return 0;
    return S_ISDIR(buffer.st_mode);
}

// 輔助函數：讀取文件全部內容
char *read_file(const char *filename, size_t *len) {
    FILE *fp = fopen(filename, "rb");
    if (!fp) return NULL;
    fseek(fp, 0, SEEK_END);
    *len = ftell(fp);
    rewind(fp);
    char *data = malloc(*len + 1);
    if (!data) { fclose(fp); return NULL; }
    fread(data, 1, *len, fp);
    data[*len] = '\0';
    fclose(fp);
    return data;
}

// 輔助函數：寫入文件
int write_file(const char *filename, const char *data, size_t len) {
    FILE *fp = fopen(filename, "wb");
    if (!fp) return -1;
    fwrite(data, 1, len, fp);
    fclose(fp);
    return 0;
}

// FNV-1a 哈希函數
unsigned int fnv1a_hash(const char *data, size_t len) {
    unsigned int hash = FNV1A_OFFSET;
    for (size_t i = 0; i < len; i++) {
        hash ^= (unsigned char)data[i];
        hash *= FNV1A_PRIME;
    }
    return hash;
}

// 將哈希值轉為 8 位十六進制字符串
void hash_to_hex(unsigned int hash, char *hex_str) {
    sprintf(hex_str, "%08x", hash);
}

// 初始化倉庫
int init_git0() {
    if (dir_exists(GIT0_DIR)) {
        printf("git0 repository already exists.\n");
        return -1;
    }
    mkdir(GIT0_DIR, 0755);
    mkdir(OBJECTS_DIR, 0755);
    fclose(fopen(INDEX_FILE, "w"));
    fclose(fopen(COMMITS_FILE, "w"));
    printf("Initialized empty git0 repository in %s/\n", GIT0_DIR);
    return 0;
}

// 添加文件到暫存區
int add_file(const char *filename) {
    if (!file_exists(filename)) { printf("File not found: %s\n", filename); return -1; }
    if (!dir_exists(GIT0_DIR)) { printf("Not a git0 repo. Run 'init' first.\n"); return -1; }

    size_t len;
    char *content = read_file(filename, &len);
    unsigned int hash = fnv1a_hash(content, len);
    char hash_hex[HASH_LEN + 1];
    hash_to_hex(hash, hash_hex);

    char obj_path[256];
    snprintf(obj_path, sizeof(obj_path), "%s/%s", OBJECTS_DIR, hash_hex);
    if (!file_exists(obj_path)) write_file(obj_path, content, len);
    free(content);

    size_t index_len;
    char *index_content = read_file(INDEX_FILE, &index_len);
    char new_index[4096] = {0};
    int found = 0;

    if (index_content) {
        char *line = strtok(index_content, "\n");
        while (line) {
            char f[256], h[HASH_LEN + 1];
            sscanf(line, "%255s %8s", f, h);
            if (strcmp(f, filename) == 0) {
                sprintf(new_index + strlen(new_index), "%s %s\n", filename, hash_hex);
                found = 1;
            } else {
                sprintf(new_index + strlen(new_index), "%s\n", line);
            }
            line = strtok(NULL, "\n");
        }
        free(index_content);
    }
    if (!found) sprintf(new_index + strlen(new_index), "%s %s\n", filename, hash_hex);
    write_file(INDEX_FILE, new_index, strlen(new_index));

    printf("Added %s to index.\n", filename);
    return 0;
}

// 提交變更
int commit_changes(const char *message) {
    if (!dir_exists(GIT0_DIR)) { printf("Not a git0 repo. Run 'init' first.\n"); return -1; }

    size_t index_len;
    char *index_content = read_file(INDEX_FILE, &index_len);
    if (!index_content || index_len == 0) {
        printf("No changes to commit. Use 'add' first.\n");
        if (index_content) free(index_content);
        return -1;
    }

    char parent_id[HASH_LEN + 1] = "none";
    size_t commits_len;
    char *commits_content = read_file(COMMITS_FILE, &commits_len);
    if (commits_content && commits_len > 0) {
        // --- 修復 2：這裡現在可以正常調用了 ---
        char *last = strrstr(commits_content, "commit ");
        if (last) sscanf(last, "commit %8s", parent_id);
        free(commits_content);
    }

    srand(time(NULL));
    unsigned int commit_hash = (unsigned int)time(NULL) ^ (rand() << 16);
    char commit_id[HASH_LEN + 1];
    hash_to_hex(commit_hash, commit_id);

    time_t now = time(NULL);
    char timestamp[64];
    strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", localtime(&now));

    char commit_content[8192];
    sprintf(commit_content,
        "commit %s\nparent %s\ndate %s\nmessage %s\nfiles\n%s\n",
        commit_id, parent_id, timestamp, message, index_content);
    free(index_content);

    FILE *fp = fopen(COMMITS_FILE, "a");
    fputs(commit_content, fp);
    fclose(fp);

    printf("Committed as %s\n", commit_id);
    return 0;
}

// 輔助：查找字符串最後一次出現的位置 (實現部分移到後面不要緊，因為前面已經宣告過了)
char *strrstr(const char *haystack, const char *needle) {
    char *last = NULL, *cur = strstr(haystack, needle);
    while (cur) { last = cur; cur = strstr(cur + 1, needle); }
    return last;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: git0 <command> [args]\n");
        printf("  init          Initialize repo\n");
        printf("  add <file>    Add file to index\n");
        printf("  commit <msg>  Commit changes\n");
        return 0;
    }
    if (strcmp(argv[1], "init") == 0) return init_git0();
    if (strcmp(argv[1], "add") == 0) return argc >=3 ? add_file(argv[2]) : (printf("Usage: add <file>\n"), -1);
    if (strcmp(argv[1], "commit") == 0) return argc >=3 ? commit_changes(argv[2]) : (printf("Usage: commit <msg>\n"), -1);
    printf("Unknown command: %s\n", argv[1]);
    return -1;
}

File: code/系統程式/textdata/git0/git0.rs
use std::env;
use std::fs::{self, OpenOptions};
use std::io::{self, Write};
use std::path::Path;
use std::process;
use std::time::{SystemTime, UNIX_EPOCH};

const GIT0_DIR: &str = ".git0";
const OBJECTS_DIR: &str = ".git0/objects";
const INDEX_FILE: &str = ".git0/index";
const COMMITS_FILE: &str = ".git0/commits";
const FNV1A_OFFSET: u32 = 0x811C9DC5;
const FNV1A_PRIME: u32 = 0x01000193;

// FNV-1a 雜湊函數
fn fnv1a_hash(data: &[u8]) -> u32 {
    let mut hash = FNV1A_OFFSET;
    for &byte in data {
        hash ^= byte as u32;
        // 使用 wrapping_mul 防止溢位報錯，等同於 C 語言預設的溢位行為
        hash = hash.wrapping_mul(FNV1A_PRIME);
    }
    hash
}

// 初始化倉庫
fn init_git0() -> io::Result<()> {
    if Path::new(GIT0_DIR).exists() {
        println!("git0 repository already exists.");
        return Ok(());
    }
    
    fs::create_dir(GIT0_DIR)?;
    fs::create_dir(OBJECTS_DIR)?;
    fs::File::create(INDEX_FILE)?;
    fs::File::create(COMMITS_FILE)?;
    
    println!("Initialized empty git0 repository in {}/", GIT0_DIR);
    Ok(())
}

// 添加文件到暫存區
fn add_file(filename: &str) -> io::Result<()> {
    if !Path::new(filename).exists() {
        println!("File not found: {}", filename);
        return Ok(());
    }
    if !Path::new(GIT0_DIR).is_dir() {
        println!("Not a git0 repo. Run 'init' first.");
        return Ok(());
    }

    // 讀取檔案並計算雜湊
    let content = fs::read(filename)?;
    let hash = fnv1a_hash(&content);
    let hash_hex = format!("{:08x}", hash);

    // 寫入 objects
    let obj_path = format!("{}/{}", OBJECTS_DIR, hash_hex);
    if !Path::new(&obj_path).exists() {
        fs::write(obj_path, &content)?;
    }

    // 更新 index 檔案
    let index_content = fs::read_to_string(INDEX_FILE).unwrap_or_default();
    let mut new_index = String::new();
    let mut found = false;

    for line in index_content.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 && parts[0] == filename {
            new_index.push_str(&format!("{} {}\n", filename, hash_hex));
            found = true;
        } else {
            new_index.push_str(&format!("{}\n", line));
        }
    }

    if !found {
        new_index.push_str(&format!("{} {}\n", filename, hash_hex));
    }

    fs::write(INDEX_FILE, new_index)?;
    println!("Added {} to index.", filename);
    
    Ok(())
}

// 提交變更
fn commit_changes(message: &str) -> io::Result<()> {
    if !Path::new(GIT0_DIR).is_dir() {
        println!("Not a git0 repo. Run 'init' first.");
        return Ok(());
    }

    let index_content = fs::read_to_string(INDEX_FILE).unwrap_or_default();
    if index_content.trim().is_empty() {
        println!("No changes to commit. Use 'add' first.");
        return Ok(());
    }

    let commits_content = fs::read_to_string(COMMITS_FILE).unwrap_or_default();
    let mut parent_id = String::from("none");

    // 尋找最後一次 commit (等同於 C 語言中的 strrstr)
    if let Some(last_idx) = commits_content.rfind("commit ") {
        let substring = &commits_content[last_idx..];
        let parts: Vec<&str> = substring.split_whitespace().collect();
        if parts.len() >= 2 {
            parent_id = parts[1].to_string();
        }
    }

    // 產生 Commit Hash (為保持無第三方依賴，使用系統時間戳 ^ 進程 ID 來模擬亂數)
    let sys_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    let commit_hash = (sys_time.as_secs() as u32) ^ (process::id() << 16);
    let commit_id = format!("{:08x}", commit_hash);

    // 時間戳格式 (為保持無第三方依賴，純使用 UNIX 時間戳)
    let timestamp = format!("Epoch {}", sys_time.as_secs());

    let commit_content = format!(
        "commit {}\nparent {}\ndate {}\nmessage {}\nfiles\n{}\n",
        commit_id, parent_id, timestamp, message, index_content
    );

    // 附加模式寫入
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(COMMITS_FILE)?;
    
    file.write_all(commit_content.as_bytes())?;

    println!("Committed as {}", commit_id);
    Ok(())
}

fn main() {
    let args: Vec<String> = env::args().collect();
    
    if args.len() < 2 {
        println!("Usage: git0 <command> [args]");
        println!("  init          Initialize repo");
        println!("  add <file>    Add file to index");
        println!("  commit <msg>  Commit changes");
        process::exit(0);
    }

    let command = args[1].as_str();
    
    let result = match command {
        "init" => init_git0(),
        "add" => {
            if args.len() >= 3 {
                add_file(&args[2])
            } else {
                println!("Usage: add <file>");
                process::exit(1);
            }
        }
        "commit" => {
            if args.len() >= 3 {
                commit_changes(&args[2])
            } else {
                println!("Usage: commit <msg>");
                process::exit(1);
            }
        }
        _ => {
            println!("Unknown command: {}", command);
            process::exit(1);
        }
    };

    if let Err(e) = result {
        eprintln!("Error executed command: {}", e);
        process::exit(1);
    }
}